infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvironmentVariablesSorter.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.provision.env;
+
+import static java.lang.String.format;
+import static java.util.Comparator.naturalOrder;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.fabric8.kubernetes.api.model.EnvVar;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+
+/**
+ * If there are any references to other environment variables in the value of some environment
+ * variable, the "dependent" environment variable needs to be defined "later" than the env var it
+ * depends on for Kubernetes variable expansion to work.
+ *
+ * <p>This class makes sure that the environment variables are sorted.
+ */
+final class EnvironmentVariablesSorter implements Comparator<EnvVar> {
```

Comment:
This is specific to how we need to sort the environment variables to kubernetes. This class is meant to merely be a helper to the `EnvVarsConverter` which does the meaningful work - applies the list of env vars to k8s.

Now I didn't make this a generic tool for sorting environment variables because the variable reference syntax is Kubernetes specific and hence only makes sense in k8s-based infrastructures. If some unknown future infrastructure is based on kubernetes, it can reuse the converter, which will do the right thing.

Generalizing the variable reference syntax seemed premature to me.

Reply:
I didn't mean generalizing the syntax or purpose, just making class public. The case I was thinking of is when some another component would want to add/use some env vars in kubernetes. But I agree that since we don't have such a case at the moment it can be premature to make it public